### PR TITLE
Custom ESLint plugin to Skip Unit Tests if Preceded by Comment with Github Issue Link 

### LINKED ISSUE #2573: Custom ESLint plugin to allow skipped unit tests only if there is a comment attached to the test with a GitHub issue link
## Problem

`eslint-plugin-jest` and `eslint-plugin-playwright` disallow disabled tests. However, we use disabled tests as part of our flaky test workflow. In that case, we require a comment is added to the test. In these cases, we are required to add the issue comment _and_ add an `eslint-disable-next-line` pragma to silence the ESLint rule disallowing disabled tests.

## Description

It would be nice to have a smarter version of the `no-disabled-test` rule that checks whether a comment immediately precedes the skipped test with an issue comment link and only raises an error if none exists.

Once this rule exists, disable the `no-disabled-test` rules from the playwright and jest plugins.

## Additional context

I thought I had worked on something like this for another project in the past but I cannot remember where any more nor find any plugins that do this for you already. We'd have to write the rule from scratch. This is possible without much fuss now that #2482 is merged.

### PR DESCRIPTION
**Pull request by chenikabukes**
## Fixes

Fixes #2573 by @sarayourfriend 

## Description

The PR adds a custom no-disabled-test rule and overrides the no-disabled-test rules in jest and playwright.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [NA] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by chenikabukes** [comment]
Note: unable to edit labels to resolve PR label check

**Comment by AetherUnbound** [comment]
> Note: unable to edit labels to resolve PR label check

No worries, when a PR is created with the pull request template we have (and references an issue the PR is intended to fix), we have automation which pulls over the labels from the issue in question and applies them to the PR (which typically resolves the PR label check steps). I've kicked that off again for this PR, so you should be set there!

**Review by sarayourfriend — CHANGES_REQUESTED** [review]
@chenikabukes This is a great start, but I have three requests for changes:

1. Please put this code into the `@openverse/eslint-plugin` package in `packages/eslint-plugin` along with unit tests for valid and invalid code.
2. The plugin must handle all `it, `describe`, and `test` with skips as well as `it.each.skip` and so forth. Please refer to the Jest and Playwright API documentation and search for "skip" on the page to get the full list of cases the plugin must match: https://jestjs.io/docs/api and https://playwright.dev/docs/api/class-test
3. The plugin should check specifically that the comment includes a link to GitHub issue, not just that a comment exists.

**Comment by chenikabukes** [comment]
> @chenikabukes This is a great start, but I have three requests for changes:
> 
> 1. Please put this code into the `@openverse/eslint-plugin` package in `packages/eslint-plugin` along with unit tests for valid and invalid code.
> 2. The plugin must handle all `it, `describe`, and `test`with skips as well as`it.each.skip` and so forth. Please refer to the Jest and Playwright API documentation and search for "skip" on the page to get the full list of cases the plugin must match: https://jestjs.io/docs/api and https://playwright.dev/docs/api/class-test
> 3. The plugin should check specifically that the comment includes a link to GitHub issue, not just that a comment exists.

I have moved the no-disabled-test rule to packages/ eslint-plugins. 

The new no-disabled-test rule has modified regex to include all types of skipped tests by playwright and jest. As well, the rule enforces the need for a GitHub issues link in a preceding comment to the skipped test. 

I have written valid and invalid tests for the new rule under packages/eslint-plugins.

**Comment by sarayourfriend** [comment]
> **Note: failing linting on the new test file because of trailing whitespace. When ESLint is run locally then this issue does not appear so I was not able to locate and thus address the issue (any tips appreciated).

@chenikabukes Can you confirm that you're running `just lint` to run the linting locally? We use `pre-commit` to run prettier in this repository which is generally the hook that handles whitespace issues.

**Comment by sarayourfriend** [comment]
Thanks for moving the rule and updating the regex to include the different APIs. I will try to test this locally. I'm surprised a single regex is able to cover all the cases, kudos! There are a few more things that need to be done before the rule can actually run locally, right now it's not doing anything:

1. Please convert the rule and its tests to TypeScript and match the implementation style of the existing rule in the `eslint-plugin` package (JavaScript files are entirely ignored in that package)
2. Add the rule to the exported rules in `rules.ts`: https://github.com//WordPress/openverse/blob/b2264d417a3336531b1ec52eaa88465a7126b123/packages/eslint-plugin/src/rules/index.ts#L3-L5
3. Update the `recommended` configuration for the plugin to include the new rule: https://github.com//WordPress/openverse/blob/53c2e70171e22593027b184b599c530a47429a03/packages/eslint-plugin/src/configs/recommended.ts#L1-L10
4. Add a documentation page for the rule, matching the style of the analytics configuration rule documentation: https://github.com//WordPress/openverse/blob/b2264d417a3336531b1ec52eaa88465a7126b123/documentation/packages/eslint_plugin/analytics_configuration.md

Once these are resolved we will be able to run and test the rule locally with the actual code base.

**Comment by chenikabukes** [comment]
Hi @sarayourfriend I've made the requested changes :) When I run **npm run test:unit** in the packages/es-lint folder then all my test cases pass. However, when I run **just lint** then one error coming from my test file no-disabled-test.spec.ts arises. Might you have an idea of where this discrepancy is coming from?

**Comment by sarayourfriend** [comment]
Just a heads-up to use `pnpm` instead of `npm` for Openverse. To run the package's tests you can do `pnpm --filter eslint-plugin run test:unit`, for example. Might be a little easier and will ensure configurations are working correctly.

There isn't any discrepancy between `test:unit` passing and `just lint` not passing. Our `pre-commit` configuration includes various formatting tools that will auto-fix or warn you about issues in files. Our unit testing scripts only run unit tests, nothing else. Anything `just lint` warns about you should fix and commit. Anything it auto-fixes for you (the Prettier and ESLint pre-commit configurations both auto-fix), you should also commit.

**Comment by chenikabukes** [comment]
@sarayourfriend Thanks for the quick reply! Noted - so with regards to the failing lint case, given the output below:
image
Why would the test case error message be giving such a problem?

**Comment by sarayourfriend** [comment]
Ahh! I wonder if it's just matching the text of the ESLint test case? I suppose this is a distinct disadvantage to using a regex checking the raw source code rather than actually parsing the call expression members/parents etc., that it will match things that aren't real function calls. Feel free to put a `// eslint-disable-next-line @openverse/no-disabled-test` to prevent the error if it's matching one of the invalid code samples in the test cases, but ideally the rule would only match real function calls, not anything inside a string or comment, for example.

**Comment by sarayourfriend** [comment]
@chenikabukes Just converting this to a draft to prevent reminder pings. Please reach out if you need any assistance with the PR. I'll be on holiday for the next two weeks or so but anyone from [`@WordPress/openverse-frontend`](https://docs.openverse.org/meta/communication_aliases.html) should be able to assist in the meantime. Once you're ready for the PR to be reviewed again, please mark it as ready for review :slightly_smiling_face:

File: packages/eslint-plugin/test/rules/no-disabled-test.spec.ts
Comment (sarayourfriend) [inline packages/eslint-plugin/test/rules/no-disabled-test.spec.ts:91]: "multi-line" comments in JavaScript are usually the `/** */` style, at least when I think of "multi-line", and [MDN refers to them as multi-line as well](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Grammar_and_types#comments).

If you instead mean multiple comments more generally, can you change this (and the other case of this in the docs) to say "multiple comments"? If the rule does not support `/** */`, that should also be documented. Ideally it would support them, though.

File: .eslintrc.js
Comment (sarayourfriend) [inline .eslintrc.js:199]: The disabled rules need to be prefixed with the plugin name. `jest` in this case, `playwright` in the other. I've suggested adding a comment to help explain why the rule is disabled.

```suggestion
        // Superceeded by `@openverse/no-disabled-test`
        "jest/no-disabled-test": "off",
```

File: .eslintrc.js
Comment (sarayourfriend) [inline .eslintrc.js:226]:
```suggestion
        // Superceeded by `@openverse/no-disabled-test`
        "playwright/no-disabled-test": "off",
```

**Review by sarayourfriend — CHANGES_REQUESTED** [review]
This is looking nearly perfect now, thank you very much for working so diligently on this contribution. I've left a few very minor comments and suggestions that will get this over the finish line and merged :slightly_smiling_face:

File: documentation/packages/eslint_plugin/no-disabled-test.md
Comment (chenikabukes) [reply documentation/packages/eslint_plugin/no-disabled-test.md:50]: It's in the invalidTestCases object :)

**Review by sarayourfriend — APPROVED** [review]
Thank you so much @chenikabukes for the patience and hard work on this contribution :pray: 

I pushed some very small, minor changes, including a rename of the test to something more explicit. After testing it locally, I realised that if all someone saw was the rule name, without the error message, it would be easy to think that the rule should be skipped, like we used to do for the others!

Anyway, thanks again, this should be good to merge once another maintainer reviews it in the coming days :tada:

**Review by dhruvkb — APPROVED** [review]
This looks fantastic, and will be a great addition to the ESLint plugin! Thank you @chenikabukes for contributing to this and addressing all the feedback.

Also thank you @sarayourfriend for the regular reviews and the changes to cross it over the finish line.

**Pull request merged**
